subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/CacheBackedTaskHistoryRepository.java

```diff
@@ -129,10 +129,13 @@ public void persist() {
                 LazyTaskExecution previousExecution = getPreviousExecution();
 
                 currentExecution.storeSnapshots();
+                taskHistoryCache.put(task.getPath(), currentExecution.snapshot());
+                // Important: we only remove unnecessary snapshots after the task history has been persisted.
```

Comment:
Good catch. The comment tells me that this API is confusing though. Removing unnecessary stuff from the *previous* execution should intuitively not affect the *current* execution. It would be great if you could find a way to slightly refactor this so it is clear and obvious without the comment.

Reply:
I refactored. Please have a look if you find it cleaner.